Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/NetworkedMessageChannel.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using Unity.Collections;
+using Unity.Netcode;
+using UnityEngine;
+
+namespace Unity.Multiplayer.Samples.BossRoom.Shared.Infrastructure
+{
+    /// <summary>
+    /// This type of message channel allows the server to publish a message that will be sent to clients as well as
+    /// being published locally. Clients and the server both can subscribe to it, but it needs to be done after
+    /// the NetworkManager is initialized.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class NetworkedMessageChannel<T> : MessageChannel<T> where T : unmanaged
+    {
+        string m_Name;
+
+        int m_BufferSize;
+
+        bool m_HasRegisteredHandler;
+
+        public NetworkedMessageChannel(int bufferSize)
+        {
+            m_Name = $"{nameof(T)}NetworkMessageChannel";
+            m_BufferSize = bufferSize;
+        }
+
+        ~NetworkedMessageChannel()
+        {
+            if (NetworkManager.Singleton != null && NetworkManager.Singleton.CustomMessagingManager != null)
+            {
+                NetworkManager.Singleton.CustomMessagingManager.UnregisterNamedMessageHandler(m_Name);
+            }
+        }
+
+        public override IDisposable Subscribe(Action<T> handler)
+        {
+            if (NetworkManager.Singleton != null && NetworkManager.Singleton.IsListening)
+            {
+                if (!m_HasRegisteredHandler && NetworkManager.Singleton.IsClient)
+                {
+                    NetworkManager.Singleton.CustomMessagingManager.RegisterNamedMessageHandler(m_Name, ReceiveMessageThroughNetwork);
+                    m_HasRegisteredHandler = true;
+                    NetworkManager.Singleton.OnClientDisconnectCallback += OnClientDisconnect;
+                }
+
+                return base.Subscribe(handler);
+            }
+
+            Debug.LogError("Cannot subscribe to NetworkedMessageChannel. NetworkManager is not initialized.");
+            return null;
+        }
+
+        void OnClientDisconnect(ulong clientId)
```

Comment:
This gets fired on the server for any client disconnect, so you may need to check if the clientId matches the LocalClientId before unregistering.

Reply:
good point!

Reply:
Another thing with this is that if a subscriber has a lifetime bigger than a networked session, it needs to make sure to unsubscribe when disconnecting and re-subscribe if connecting to a new session. I'll add something to that effect in the class summary